drivers: initial commit of generic AT parser module

kaspar030 opened the pull request:
This PR adds a module for interacting with devices using AT commands.

miri64 commented:
@jia200x maybe useful for your https://github.com/RIOT-OS/RIOT/pull/5427

vincent-d commented:
I started to work on prototyping with our modem some weeks ago, so I gave a try to your code.

It worked good once I understood I had to add a `\r` to each of my commands ;)

This is much more simpler that the system I used on a previous project which worked with a lot of callbacks. But here it's hard to deal with responses you do not expect.

I'll give some more try!

jia200x commented:
@miri64 Indeed. Thanks for pointing me to this.

Cheers

kaspar030 commented:
> It worked good once I understood I had to add a \r to each of my commands ;)

oh, I hoped I had found a combination of line endings to send / line endings to expect that works. The two modems I tried accept a line with just ```\n```, but echo ```\r\n```. They don't like to receive crlf, though. Maybe we need to parameterize the newline behaviour somehow?

kaspar030 commented:
> This is much more simpler that the system I used on a previous project which worked with a lot of callbacks. But here it's hard to deal with responses you do not expect.

Very good point. I planned to add support for that later, once I come up with a process that works and doesn't suck. Suggestions welcome!

vincent-d commented:
> Maybe we need to parameterize the newline behaviour somehow?

It would be handy, indeed.

I use a ublox modem, which needs a `\r` and echoes `\r\r\n`.

aabadie commented:
Any update here ? @kaspar030

vincent-d commented:
I gave another try and proposed some improvements in https://github.com/kaspar030/RIOT/pull/27

haukepetersen commented on drivers/at/at.c:
```diff
@@ -0,0 +1,238 @@
+        return -1;
+    }
+
+    if (at_expect_bytes(dev, AT_END_OF_LINE "\r\n", sizeof(AT_END_OF_LINE) + 1, timeout)) {
```
shouldn't this be `AT_END_OF_LINE "\n"` as second parameter?

vincent-d replied:
On my modem (and I think it works the same for others as well), if you send `AT\r` you get as a response:
`AT\r` <- echo
`\r\nOK\r\n`

haukepetersen replied:
maybe I missunderstand the line above, but `AT_END_OF_LINE` is
```c
#define AT_END_OF_LINE "\r"
```
So the line above does make a string concatenation, leading to `\r\r\n`, right?!

vincent-d replied:
Yes

haukepetersen replied:
And this is wanted? I don't quite see it, but if it is intended then never mind...

vincent-d replied:
Yes, the `AT_END_OF_LINE` is expected after the command, the `\r\n` before the first response line. 

It seems that some modems can use different end of line character than `\r`, that's why it's configurable.

haukepetersen replied:
alright, I see.

haukepetersen replied:
but just to make sure: wouldn't it make sense then, to make the characters that are expected before the first response line also configurable?

vincent-d replied:
That would not hurt, indeed

aabadie commented on drivers/include/at.h:
```diff
@@ -0,0 +1,202 @@
+ * @param[in]   len         len of @p buffer
+ * @param[in]   timeout     timeout (in usec)
+ *
+ * @returns     lenght of response on success
```
before I forget, there's a typo: s/lenght/length/. I saw it in other places.

kaspar030 replied:
fixed two times int at.h.

keestux commented:
@aabadie Sorry, I don't have a LoRaBee at hand. I _do_ have sodaq ONE and ExpLoRer board with the same RN2483 device but unfortunately I can't connect my Atmel ICE to these boards (so I can't load RIOT programs). :-(

keestux commented on drivers/include/at.h:
```diff
@@ -0,0 +1,202 @@
+ *
+ * @param[in]   dev     device to operate on
+ * @param[in]   bytes   buffer containing bytes to expect
+ * @param[in]   len     number of bytes to expect
```
Why do we need 'len'? The length of 'bytes' should be leading, otherwise you need a sanity check in the function.

kaspar030 replied:
I don't quite get the question. ```len``` specifies the nr of bytes in ```bytes```. Maybe the docs are misleading?

keestux replied:
Why not rely on strlen(bytes) instead of having to pass a string and a length? In the implementation there is a while loop for 'len', but there is no sanity check whether there are that many chars in 'bytes'.

Without 'len' the usage could become as simple as:

    if (at_expect_bytes(dev, AT_END_OF_LINE "\r\n", timeout)) {

Bonus: in this case getting rid of the confusing sizeof()+1

kaspar030 replied:
> Why not rely on strlen(bytes) instead of having to pass a string and a length?

the function is mostly used by other functions (quick grep counts five times). relying on strlen would lead to counting strings multiple times, for strings that are often static. having ```len``` as parameter allows the API user to decide whether to use strlen or not...



> there is no sanity check whether there are that many chars in 'bytes'

keestux replied:
I'm not impressed by that argument.
You can change the while condition in at_expect_bytes() to

    while (*bytes != 0) {

and then you won't need to do a strlen()

kaspar030 replied:
That actually seems like a good idea. done.

keestux commented on drivers/at/at.c:
```diff
@@ -0,0 +1,238 @@
+
+    memset(resp_buf, 0, len);
+
+    while (len) {
```
What happens if the response string is longer than 'len' and the while loop is ended. If I understand this code well enough then I'm afraid there is no NUL byte terminating the string.

kaspar030 replied:
That's what the memset in line 210 should take care of?

keestux replied:
No, not quite. Say there is a 'len' of 5 and on the input comes "abcdefg\r\n". The while loop ends as soon as "abcde" was read and resp_buf is completely filled up (no NUL byte anymore). No?

kaspar030 replied:
Ah, right. I guess ```-ENOSPC``` would be the desired result in that case?

kaspar030 replied:
Looking at the code again, if no full line could be read, the result of the function is -1. Shouldn't that be enough? Anyhow, I've added a condition setting the first byte of resp_buf to the null byte in that case.

keestux replied:
I still don't see how the string is terminated with a nul-byte. In my example above resp_buf contains the 5 characters a, b, c, d and e. The return value is 5. Maybe you say that the caller is responsible to not treat the result as a nul-terminated string?

kaspar030 replied:
The return variable ```res```, which defaults to ```-1```, is only set to the number of bytes read if a full line could be read. In your example (```[ a, b, c, d, e ]```), the while loop would end with ```res == -1``` and thus set ```*resp_buf = '\0'```.

kaspar030 replied:
@keestux Does this address your concerns?

keestux replied:
I can see that there can be no unterminated string, even when a longer line comes in. Good.

On the other hand, you take drastic measures to throw away the whole input (instead of truncating).

If this is what we want then OK.
BTW. The doc says "<0 on error", which is either a timeout (-ETIMEDOUT) or a too long input line (-1). Perhaps you want to put this in the function description (at.h).

smlng commented on drivers/include/at.h:
```diff
@@ -0,0 +1,202 @@
+ * Most functions compare the bytes echoed by the device with what they
+ * intended to send, and bail out if there's no match.
+ *
+ * Furthermore, the library tries to copy with difficulties regarding different
```
/copy/cope/ ?

kaspar030 replied:
fixed

kaspar030 commented:
> Do you think it could be even more generalized to any string command controlled devices ? (not only AT).

It probably can even be used as is for devices not speaking Hayes. The actual AT commands are all built on top of ```at_expect_bytes()```, ```at_readline()``` and ```at_send_bytes()```.

kaspar030 commented:
Can we get forward with this? *ping* @haukepetersen @aabadie @keestux

aabadie commented on drivers/Makefile.dep:
```diff
@@ -4,6 +4,11 @@ ifneq (,$(filter adxl345,$(USEMODULE)))
 endif
 
+ifneq (,$(filter at,$(USEMODULE)))
+    USEMODULE += fmt
```
indent should be 2 spaces

kaspar030 replied:
fixed

aabadie commented:
I'll need this PR soon for a new radio driver

aabadie commented on drivers/include/at.h:
```diff
@@ -0,0 +1,200 @@
+ * @param[in]   bufsize     size of @p buf
+ *
+ * @returns     0 on success
+ * @returns     <0 otherwise
```
Using an anonymous enum for return codes would improve the API of this library.

kaspar030 replied:
not in this PR

aabadie commented on drivers/include/at.h:
```diff
@@ -0,0 +1,200 @@
+ * @param[in]   bufsize     size of @p buf
+ *
+ * @returns     0 on success
+ * @returns     <0 otherwise
```
Using an anonymous enum for return codes would improve the API of this library.

aabadie commented on drivers/include/at.h:
```diff
@@ -0,0 +1,200 @@
+/**
+ * @brief   send AT command, wait for a prompt
+ *
+ * This function will send the supplied @p command, then wait for the prompt (>)
```
"This functions sends ..., then waits for ...". Prefer present instead of future (you did it like this above). Same in the documentations below.

kaspar030 replied:
done and fixed some more style issues in the doxygen

kaspar030 commented on drivers/at/at.c:
```diff
@@ -0,0 +1,240 @@
+            if (AT_PRINT_INCOMING) {
+                print(&c, 1);
+            }
+            if (!(c == *bytes++)) {
```
yup

kaspar030 commented on drivers/at/at.c:
```diff
@@ -0,0 +1,240 @@
+    return res;
+}
+
+ssize_t at_send_cmd_get_lines(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout)
```
done

kaspar030 commented on drivers/at/at.c:
```diff
@@ -0,0 +1,240 @@
+        goto out;
+    }
+
+    while(1) {
```
yup

kaspar030 commented:
- rebased
- addressed @aabadie's comments
- renamed ```AT_END_OF_LINE``` -> ```AT_EOL```, added ```AT_EOL_LEN``` define

vincent-d commented:
Could we move forward on this one? I have some work pending waiting for this. @aabadie @keestux @haukepetersen ?

keestux commented:
Since you asked ... I went over `at.c/at.h` once more and have the following findings. Sorry for the long message.

- Some devices (for example UBlox Cellular Modules, and SIM800/SIM900) have "Unsolicited Response Codes". These URC's can come in at any moment. That would cause problems if that happens between `uart_write` and `at_expect_bytes.` So, it is debatable whether the `at_expect_bytes` calls in `at_send_cmd` are a good idea. I understand that we want the at_expect_bytes to tightly synchronize the character we send with the reply from the modem. But maybe we need to do that character by character.

- The timeout in `at_drain` a hard coded number. It deserves a comment explaining why it is 10ms. No?

- Why use `unsigned` instead of `size_t?` (2x)

- Besides `+CME ERROR`, there is also `+CMS ERROR`. (In UBlox and SIM, for SMS errors)

- In `at_send_cmd_get_lines` the `out` label is used only once. Either do a simple return here, or do `goto out` in the remainder of the function.

- In `at_readline` there is a `goto out`, which could easily be replaced by a `break`, and then drop the label `out`.

- In `at_send_cmd_wait_prompt` there almost a complete repetition of `at_send_cmd`. Why is that? The only difference is the following. (If that's intended, please add a comment. If not, please replace by a call to `at_send_cmd`)

This

    if (at_expect_bytes(dev, AT_EOL "\r\n", timeout)) {

versus this

    if (at_expect_bytes(dev, AT_EOL "\n", timeout)) {

- A remark about the comment in at.h

    It usually sends "< command>< CR>", but expects "< command>\LF\CR" as echo.

- Notice that in `at_send_cmd` the code is expecting "< command>< CR>< CR>< LF>" (see also my remark about code duplication). Perhaps we need two macros, one for the < CR> to send a command and one for < CR>< LF> as the end-of-line.

BTW. I did all this just by code inspection, using my knowledge of writing the GPRSbee and Sodaq_3Gbee Arduino libraries. I did not actually test anything with RIOT, sorry.

kaspar030 commented:
> Some devices (for example UBlox Cellular Modules, and SIM800/SIM900) have "Unsolicited Response Codes". These URC's can come in at any moment. That would cause problems if that happens between uart_write and at_expect_bytes. So, it is debatable whether the at_expect_bytes calls in at_send_cmd are a good idea. I understand that we want the at_expect_bytes to tightly synchronize the character we send with the reply from the modem. But maybe we need to do that character by character.

URCs cannot come at any moment, AFAIK, they cannot arrive after AT has been sent and the corresponding command has been finished?

There's #8542 which might be a direction to go.
My hunch is that we need to have a different approach to URCs. Anyhow, I suggest getting this PR merged and improving later.

kaspar030 commented:
> The timeout in at_drain a hard coded number. It deserves a comment explaining why it is 10ms. No?

Fixed.

> Why use unsigned instead of size_t? (2x)

In at_send_cmd() and it's prompting brother? fixed

> Besides +CME ERROR, there is also +CMS ERROR. (In UBlox and SIM, for SMS errors)

Added.

> In at_send_cmd_get_lines the out label is used only once. Either do a simple return here, or do goto out in the remainder of the function.
> 
> In at_readline there is a goto out, which could easily be replaced by a break, and then drop the label out.

Do those once-used labels hurt a lot?

> In at_send_cmd_wait_prompt there almost a complete repetition of at_send_cmd. Why is that? The only difference is the following. (If that's intended, please add a comment. If not, please replace by a call to at_send_cmd)

I merged a PR to this PR, which introduced the extra function. Should've seen.
Anyhow, I rewrote the function so it now calls at_send_cmd(). @vincent-d does it still work?

kaspar030 commented:
Honestly I don't want to spend too much time on this. I suggest we either merge it (possibly marking it as "experimental"), or I keep it in a seperate repo.

vincent-d commented on drivers/at/at.c:
```diff
@@ -0,0 +1,238 @@
+{
+    at_drain(dev);
+
+    int res = at_send_cmd(dev, command, timeout);
```
I didn't re-test yet, but IIRC, this won't work anymore.
Because in case of the prompt, the "\r\n" is not sent back.

In 'normal' case: we send `command\r\n` and we receive `command\r\n\r\nOK\r\n`. In case of a prompt, if we send `command\r\n` we receive `command\r\n>`. That's why I added a new function instead of reusing `at_send_cmd`.

keestux replied:
I thought we just send `command\r` (without the `\n`)

kaspar030 replied:
confusing at least ;)
Anyhow, I reverted to the working version. Let's optimize later.

vincent-d commented:
> does it still work?

See my inline response

>  suggest we either merge it (possibly marking it as "experimental")

+1, there is for sure room for improvements, but I'd like this to be merge to be able to move forward on related topics.

keestux commented:
I agree with @kaspar030 , let's merge it.

aabadie commented:
> I suggest we either merge it (possibly marking it as "experimental")

Like others, I also agree. From a reviewer point of view, e.g. skipping AT parser implementation details discussed above, the main problem with this PR is the missing test application. It would be great to provide a simple with a shell to send commands.
Otherwise, this module is not built by the CI.

vincent-d commented:
monthly wake-up ;)

@kaspar030 do you think you can make it for the release?

aabadie commented on tests/at/main.c:
```diff
@@ -0,0 +1,92 @@
+        puts(resp);
+    }
+    else {
+        puts("Error");
```
return 1 after ?

aabadie commented on tests/at/main.c:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    if (at_send_cmd_get_resp(&at_dev, argv[1], resp, sizeof(resp), 10 * US_PER_SEC) > 0) {
+        puts("Response:");
```
Why not use `printf("Response: %s\n", resp);` instead ? (a `\n` could be added after `:` to be strictly identical)

aabadie commented on tests/at/main.c:
```diff
@@ -0,0 +1,92 @@
+    (void)argc;
+    (void)argv;
+
+    at_dev_init(&at_dev, UART_AT, BAUDRATE_AT, buf, sizeof(buf));
```
I think, it would be more flexible to pass the UART and baudrate to the init function.

aabadie commented:
@kaspar030, if you don't mind I'll address the comments I raised.

aabadie commented on drivers/at/at.c:
```diff
@@ -0,0 +1,248 @@
+    uart_write(dev->uart, (const uint8_t *)command, cmdlen);
+    uart_write(dev->uart, (const uint8_t *)AT_SEND_EOL, AT_SEND_EOL_LEN);
+
+    if (AT_SEND_ECHO) {
```
@vincent-d @kaspar030, here is the change that allows to skip the echo check at build time.

vincent-d commented:
> What about adding a macro like AT_END_OF_RESP

I would suggest to squash and merge as-is and improve in a follow-up PR.
I also have a couple of improvements I'd like to merge later on.

@kaspar030 could you squash?

aabadie commented:
> we should fix Murdock

done! (sorry)

aabadie commented on cpu/atmega_common/periph/i2c.c:
```diff
@@ -32,9 +32,6 @@
 #include "debug.h"
 #define ENABLE_DEBUG      (0)
 
-/* guard file in case no I2C device is defined */
```
This looks unrelated, weird.

kaspar030 replied:
yes, no idea how that commit sneaked in. I squashed it out.

aabadie commented:
@vincent-d, ok to merge ? I think we can safely dismiss @haukepetersen's review

aabadie commented:
Cool !

vincent-d commented:
Great team work! Thanks all!
